Add new rule that detect retrieving environment variables from /proc files

hi120ki opened the pull request:
Signed-off-by: Hi120ki <12624257+hi120ki@users.noreply.github.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. . Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

> /kind release

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

/kind rule-create

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area engine

/area rules

> /area tests

> /area proposals

> /area CI

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

These days, it's common to pass a secret to an application running on a container via an environment variable. Secrets store very sensitive information such as cloud credentials and JWT secretkeys, so it is important to detect when environment variables have been retrieved by an attack.

And if the application has SSRF or path traversal vulnerabilities, the attacker can steal the environment variables of the application process by reading `/proc/self/environ` and `/proc/1/environ` under the `/proc` directory.

<https://linux.die.net/man/5/proc>

For example, SSRF:

```
curl -X POST -d "url=file:///proc/self/environ" https://example.com/ssrf
curl -X POST -d "url=file:///proc/1/environ" https://example.com/ssrf
```

Path traversal:

```
curl --path-as-is https://example.com/path/../../proc/self/environ
curl --path-as-is https://example.com/path/../../proc/1/environ
```

In MITRE ATT&CK, this technique is categorised as [T1552 : Unsecured Credentials](https://attack.mitre.org/techniques/T1552/), and Sub-technique [T1552.001 : Unsecured Credentials: Credentials In Files](https://attack.mitre.org/techniques/T1552/001/), and Detection is [DS0022 - File - File Access](https://attack.mitre.org/datasources/DS0022/), and component is [T1003.007 OS Credential Dumping: Proc Filesystem](https://attack.mitre.org/techniques/T1003/007/).

I propose the following as a rule to detect environment variable reading from files under the `/proc` directory.

```yaml
- list: proc_environ_file_names
  items: [/proc/self/environ, /proc/1/environ]

- macro: proc_environ_files
  condition: >
    fd.name in (proc_environ_file_names)

- rule: Read environment variable from /proc files
  desc: An attempt to read process environment variables from /proc files
  condition: >
    container and open_read and proc_environ_files
  enabled: true
  output: >
    Environment variables were retrieved from /proc files (user=%user.name user_loginuid=%user.loginuid program=%proc.name
    command=%proc.cmdline file=%fd.name parent=%proc.pname gparent=%proc.aname[2] ggparent=%proc.aname[3] gggparent=%proc.aname[4] container_id=%container.id image=%container.image.repository)
  priority: WARNING
  tags: [filesystem, mitre_credential_access, mitre_discovery]
```

> Adding a host to the detection target causes false positives by `systemd`, so limited to container.

I'm waiting for your comments on detecting reading these files and my proposed rules.

Thank you.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` or `kind/flaky-test`, please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of the rule engine`.
-->

```release-note
Add rule to detect an attempt to read process environment variables from /proc files
```

poiana commented:
@hi120ki: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign mstemm for approval by writing `/assign @mstemm` in a comment. For more information see:[The Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

poiana commented:
Welcome @hi120ki! It looks like this is your first PR to falcosecurity/falco 🎉

darryk10 commented:
Hi @hi120ki, thanks for contributing first of all :) 
Amazing rule and I agree with you on  the security use case.
I also do like the container scoping to avoid the noise since we need to rely on `open_read` events. 
Just wondering if we can open the rule to hosts, just whitelisting `systemd` as you mentioned, or you feel it might introduce too much default noise.
I'm going to test the rule as well to further reduce the noise if possible :) 
Thanks

jasondellaluce commented:
/ok-to-test

jasondellaluce commented:
/milestone 0.33.0

hi120ki commented:
@darryk10 @jasondellaluce I update my proposed rule to expand host. 
```yaml
- rule: Read environment variable from /proc files
  desc: An attempt to read process environment variables from /proc files
  condition: >
    open_read and (fd.name glob /proc/*/environ)
    and not proc.name in (systemctl, systemd-detect-, cloud-id)
  enabled: true
  output: >
    Environment variables were retrieved from /proc files (user=%user.name user_loginuid=%user.loginuid program=%proc.name
    command=%proc.cmdline file=%fd.name parent=%proc.pname gparent=%proc.aname[2] ggparent=%proc.aname[3] gggparent=%proc.aname[4] container_id=%container.id image=%container.image.repository)
  priority: WARNING
  tags: [filesystem, mitre_credential_access, mitre_discovery]
```
Host machine has many processes, and there is many `/proc/*/environ` files.(`*` is process id).
So, updated rule uses use `glob` instead of `in` and enables to detect any process's environ files.

Then, in host machine some system process such as `systemd` uses `/proc/*/environ` files, and I added allowlist into rule.
If you find other  system processes, cloud you add to allowlist?

hi120ki commented:
@darryk10 @jasondellaluce I tryed to apply my rule to my testbed, there are so many false positive detection from host machine. (For example, `systemd-tmpfile` `systemd-user-se` `systemd-update-`)
Of cource, it is nice to add all systemd and other system-level process binary to allowlist if we can, but systemd updates fast, and it is impossible to maintain this allowlist.

So, we should restrict this rule only container. I'm waiting for your comments.

darryk10 commented:
Hi @hi120ki, thanks for the further test. I've been testing the host condition and I agree the rule generates too noise. Even though would be good to have a role covering host as well, I would suggest to keep this use case for container only to reduce the noise as you suggested in the PR. 
Regarding the condition for container, instead, it's very accurate and very low false positive. In addition I observed good true positives on malwares and bad activities so definitely a rule we want in our ruleset. 
Happy to approve the PR once reverted on container use case only.
Really appreciate your work and looking forward to receiving new rules from you :)

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [6481dae](https://github.com/falcosecurity/falco/commits/6481daec4b08595c7397b749ef92099ca75275a2) Revert "add systemd-sysctl to allowlist"
- [b10fa31](https://github.com/falcosecurity/falco/commits/b10fa317be685d0bc50e40f7858ea658fa6a0de7) Revert "add host to target"

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

hi120ki commented:
@darryk10 Hi, I run this rule on my testbed, and found one false positive detection in `scsi_id` proc.
This proc runs by [gcp-compute-persistent-disk-csi-driver](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver), it is common in GKE cluster.
I added this into allowlist, could you review it?

darryk10 commented on rules/falco_rules.yaml:
```diff
@@ -3202,3 +3202,18 @@
+  condition: >
+    container and open_read and (fd.name glob /proc/*/environ)
+    and not proc.name in (known_binaries_to_read_environment_variables_from_proc_files)
+  enabled: true
```
```suggestion
```

darryk10 replied:
This isn't necessary since rules are enabled by default.

darryk10 commented:
@hi120ki I think you also need to specify the change log in the PR otherwise the PR isn't mergeable.

darryk10 reviewed (approved):
Thanks for your contribution :)

poiana commented:
LGTM label has been added.  <details>Git tree hash: 2d32729630c4cad609fc64a0f7f2b1e40e8fb977</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1101988076" title="Approved">darryk10</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*
**Once this PR has been reviewed and has the lgtm label**, please assign mstemm for approval by writing `/assign @mstemm` in a comment. For more information see:[The Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

<details open>
Needs approval from an approver in each of these files:

- **[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["mstemm"]} -->

hi120ki commented:
/assign @mstemm

jasondellaluce reviewed (dismissed):
/approve

Kaizhe commented on rules/falco_rules.yaml:
```diff
@@ -3202,3 +3202,17 @@
+- rule: Read environment variable from /proc files
+  desc: An attempt to read process environment variables from /proc files
+  condition: >
+    container and open_read and (fd.name glob /proc/*/environ)
```
to optimize rule checking performance, it is good to put `open_read` in the beginning of the rule.

hi120ki replied:
Thank you! I put open_read in the beginning of the rule.

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1102421524" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
New changes are detected. LGTM label has been removed.

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1102421524" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
LGTM label has been added.  <details>Git tree hash: 36807142c07ff0a54c78bd2f0180c69e8a4cd8ca</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1108626360" title="Approved">darryk10</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1102421524" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jasondellaluce reviewed (approved):
/approve

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1108626360" title="Approved">darryk10</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1108775193" title="Approved">jasondellaluce</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [jasondellaluce]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1108626360" title="Approved">darryk10</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#" title="Author self-approved">hi120ki</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1108775193" title="Approved">jasondellaluce</a>*, *<a href="https://github.com/falcosecurity/falco/pull/2193#pullrequestreview-1110491394" title="Approved">leogr</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalco).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[rules/OWNERS](https://github.com/falcosecurity/falco/blob/master/rules/OWNERS)~~ [jasondellaluce,leogr]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

leogr commented:
Closing and reopening to trigger the CI
/close

poiana commented:
@leogr: Closed this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/2193#issuecomment-1249221598):

>Closing and reopening to trigger the CI
>/close


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr commented:
/reopen

poiana commented:
@leogr: Reopened this PR.

<details>

In response to [this](https://github.com/falcosecurity/falco/pull/2193#issuecomment-1249221707):

>/reopen


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

leogr commented:
/hold

for the CI to complete

leogr commented:
/hold cancel
